lms/djangoapps/courseware/model_data.py

```diff
@@ -34,12 +36,644 @@ class InvalidWriteError(Exception):
+        """
+        raise NotImplementedError()
+
+
+class UserStateCache(object):
+    """
+    Cache for Scope.user_state xblock field data.
+    """
+    def __init__(self, user, course_id):
+        self._cache = defaultdict(dict)
+        self.course_id = course_id
+        self.user = user
+        self._client = DjangoXBlockUserStateClient(self.user)
+
+    def cache_fields(self, fields, xblocks, aside_types):  # pylint: disable=unused-argument
+        """
+        Load all fields specified by ``fields`` for the supplied ``xblocks``
+        and ``aside_types`` into this cache.
+
+        Arguments:
+            fields (list of str): Field names to cache.
+            xblocks (list of :class:`XBlock`): XBlocks to cache fields for.
+            aside_types (list of str): Aside types to cache fields for.
+        """
+        block_field_state = self._client.get_many(
+            self.user.username,
+            _all_usage_keys(xblocks, aside_types),
+        )
+        for usage_key, field_state in block_field_state:
+            self._cache[usage_key] = field_state
+
+    @contract(kvs_key=DjangoKeyValueStore.Key)
+    def set(self, kvs_key, value):
+        """
+        Set the specified `kvs_key` to the field value `value`.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+            value: The field value to store
+        """
+        self.set_many({kvs_key: value})
+
+    @contract(kvs_key=DjangoKeyValueStore.Key, returns="datetime|None")
+    def last_modified(self, kvs_key):
+        """
+        Return when the supplied field was changed.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: datetime if there was a modified date, or None otherwise
+        """
+        return self._client.get_mod_date(
+            self.user.username,
+            kvs_key.block_scope_id,
+            fields=[kvs_key.field_name],
+        ).get(kvs_key.field_name)
+
+    @contract(kv_dict="dict(DjangoKeyValueStore_Key: *)")
+    def set_many(self, kv_dict):
+        """
+        Set the specified fields to the supplied values.
+
+        Arguments:
+            kv_dict (dict): A dictionary mapping :class:`~DjangoKeyValueStore.Key`
+                objects to values to set.
+        """
+        pending_updates = defaultdict(dict)
+        for kvs_key, value in kv_dict.items():
+            cache_key = self._cache_key_for_kvs_key(kvs_key)
+            field_state = self._cache[cache_key]
+            field_state[kvs_key.field_name] = value
+
+            pending_updates[cache_key][kvs_key.field_name] = value
+
+        try:
+            self._client.set_many(
+                self.user.username,
+                pending_updates
+            )
+        except DatabaseError:
+            raise KeyValueMultiSaveError(sum((state.keys() for state in pending_updates.values()), []))
```

Comment:
This looks a bit cryptic. Maybe assign it to a named var to make the meaning clearer? I also find `itertools.chain` to convey this meaning more clearly, but maybe that's just a personal preference.